Animal docs update (#634)
ImperialCollegeLondon/virtual_rainforest: docs/source/virtual_ecosystem/implementation/animal_implementation.md

---
jupytext:
  formats: md:myst
  main_language: python
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
language_info:
  codemirror_mode:
    name: ipython
    version: 3
  file_extension: .py
  mimetype: text/x-python
  name: python
  nbconvert_exporter: python
  pygments_lexer: ipython3
  version: 3.11.9
---

# Animal Model Implementation

## Model Overview

The animal model in the Virtual Ecosystem platform represents the dynamics of animal
communities. The implementation is structured

> around three core classes, each playing a distinct role in managing animal agents, their

Reviewer: I would find it useful if you could mention and briefly define these classes here, no more than one sentence, and link to the respective sections

Reviewer 2: I agree. A bullet list with five word descriptions would do it, but it would be good to say here.

traits, and their interactions with the ecosystem. Together, these classes facilitate a
flexible and scalable approach to simulating complex animal processes across spatial
grids.

### Core Classes

#### **1. FunctionalGroup**

The `FunctionalGroup` class encapsulates the fixed traits of an organism, defining its
ecological role and life-history strategy. These traits include metabolic type, diet,
taxonomic classification, reproductive strategy, and developmental type. Functional
groups serve as templates for constructing animal cohorts and ensuring consistent trait
inheritance across simulations.

Key responsibilities:

- Encodes fixed traits shared by cohorts within the group.

> - Provides scaling parameters for processes like territory size and prey selection.

Reviewer 2: Are these all mass specific scaling factors? Can we unpack a bit: "Provides scaling factors that scale processes like territory size and prey selection to the mass of individuals within a cohort."

- Defines dietary and physiological constraints that influence cohort behavior.

#### **2. AnimalCohort**

> The `AnimalCohort` class represents individual animal agents at the cohort level. Each

Reviewer 2: This makes it sound like an IBM?

```suggestion
The `AnimalCohort` class represents a group of identically sized individual animal agents from a single functional group: a cohort. Each
```

cohort is age-specific, meaning all individuals in the cohort were produced in the same
reproductive event and share the same age. The class tracks dynamic state variables such
as mass, age, reproductive biomass, and territory occupancy.

Key attributes:

- **Dynamic state variables**: Mass, age, individuals, reproductive mass, and more.
- **Territory management**: Tracks the spatial extent of a cohort's interactions with
  resources and other cohorts.
- **Lifecycle processes**: Handles maturity, mortality, and metamorphosis.

#### **3. AnimalModel**

The `AnimalModel` class orchestrates the animal community at the spatial grid level.

> It contains methods for initializion, setup, and updating of the animal model as well as

Reviewer: ```suggestion
It contains methods for initialization, setup, and updating of the animal model as well as
```

communication with the core data object. Cohort-facing methods found within the animal
model are either community-level processes which loop over all cohorts in the simulation

> or are methods pertaining to the creation, destruction, or movement of animal chorts.

Reviewer: ```suggestion
or are methods pertaining to the creation, destruction, or movement of animal cohorts.
```

> communication with the core data object. Cohort-facing methods found within the animal
> model are either community-level processes which loop over all cohorts in the simulation
> or are methods pertaining to the creation, destruction, or movement of animal chorts.

Reviewer 2: ```suggestion
communication with the core data object. The animal
model provides methods that loop over all cohorts in the simulation to simulate community processes, along with methods to handle cohort movement, creation and death.
```

Key responsibilities:

- **Cohort management**: Initializes, updates, and removes animal cohorts.
- **Spatial dynamics**: Tracks cohort occupancy across grid cells and handles migration.
- **Community-level processes**: Foraging, mortality, birth, metamorphosis, and more.

### Code Structure

The `AnimalModel` is a subclass of `BaseModel`, integrating into the broader
Virtual Ecosystem framework. It extends the base functionality to include animal-specific
methods and attributes, ensuring compatibility with other ecosystem modules such as
vegetation, litter, and soil.

Key components:

> - **Initialization**: Sets up the grid, functional groups, and resource pools for
>   excrement, carcasses, and leaf waste.

Reviewer 2: ```suggestion
- **Initialization**: Sets up the grid structure for animal movement, functional groups, and resource pools for
  excrement, carcasses, and leaf waste.
```
The grid creation is part of the core.

revised in the future.

1. **Initialize Litter Pools**
   Litter pools accessible to animals for consumption are populated using the
   `populate_litter_pools` method. These pools are referenced during foraging and
   consumption calculations.

2. **Foraging**
   Animal cohorts forage for resources within their communities using the
   `forage_community` method. Resource consumption is determined by cohort traits and
   resource availability.

3. **Migration**
   Cohorts migrate between grid cells based on birth events and resource
   availability using the `migrate_community` method. Migration updates the spatial
   distribution of cohorts.

4. **Birth**
   New cohorts are generated through reproduction using the `birth_community` method.
   Parent cohorts allocate biomass from their reproductive mass to create new cohorts.

5. **Metamorphosis**
   Larval cohorts transition into adult cohorts using the `metamorphose_community`
   method. This transition updates the cohort's traits and functional role.

6. **Metabolism**
   Cohorts metabolize consumed resources, processing consumed matter into waste and

>    fueing internal processes. This is handled by the `metabolize_community` method,

Reviewer: ```suggestion
   fueling internal processes. This is handled by the `metabolize_community` method,
```
